src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs

```diff
@@ -1007,6 +1005,82 @@ static string buildAssemblyExternClause(CSharpCompilation comp)
+        {
+            var originalIdentity = new AssemblyIdentity("Ret", new Version(1, 0, 0, 0), isRetargetable: true);
+            // Custom corlib is necessary as the CallConv type must be defined in corlib, and we need to make
+            // sure that it's retargeted correctly.
+            string corlibSource = @"
+namespace System
+{
+    public class Object { }
+    public abstract class ValueType { }
+    public struct Void { }
+    public abstract partial class Enum : ValueType {}
+    public class String { }
+    public struct Boolean { }
+    public struct Int32 { }
+    public class Type { }
+    public class Attribute { }
+    public class AttributeUsageAttribute : Attribute
+    {
+        public AttributeUsageAttribute(AttributeTargets validOn) {}
+        public bool Inherited { get; set; }
+    }
+    public enum AttributeTargets { Method = 0x0040, }
+    namespace Runtime
+    {
+        namespace InteropServices
+        {
+            [AttributeUsage(AttributeTargets.Method, Inherited = false)]
+            public sealed class UnmanagedCallersOnlyAttribute : Attribute
+            {
+                public UnmanagedCallersOnlyAttribute()
+                {
+                }
+
+                public Type[] CallConvs;
+                public string EntryPoint;
+            }
+        }
+        namespace CompilerServices
+        {
+            public sealed class CallConvCdecl {}
+        }
+    }
+}
+namespace System.Runtime.InteropServices
+{
+}
+";
+            var beforeRetargeting = CreateCompilation(originalIdentity, new[] { corlibSource }, new MetadataReference[0]);
+            beforeRetargeting.VerifyDiagnostics();
+
+            var unmanagedCallersOnlyAssembly = CreateEmptyCompilation(@"
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+public class C
+{
+    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) } )]
+    public static void M(int s) {}
+}
+", new[] { beforeRetargeting.ToMetadataReference() });
+            unmanagedCallersOnlyAssembly.VerifyDiagnostics();
+
+            var afterRetargeting = CreateCompilation(originalIdentity.WithVersion(new Version(2, 0, 0, 0)), new[] { corlibSource }, new MetadataReference[0]);
+
+            var finalComp = CreateEmptyCompilation(@"", references: new[] { afterRetargeting.ToMetadataReference(), unmanagedCallersOnlyAssembly.ToMetadataReference() });
+            finalComp.VerifyDiagnostics();
+
+            var m = finalComp.GetTypeByMetadataName("C").GetMethod("M");
+            var unmanagedCallersOnlyData = m.UnmanagedCallersOnlyAttributeData;
+            Assert.IsType<RetargetingMethodSymbol>(m);
+            var containingAssembly = unmanagedCallersOnlyData.CallingConventionTypes.Single().ContainingAssembly;
+            Assert.NotSame(containingAssembly, beforeRetargeting.Assembly);
+            Assert.Same(containingAssembly, afterRetargeting.Assembly);
+        }
```

Comment:
1. Consider also verifying diagnostic on caller.
2. Do we have a retargeting scenario for empty or null CallConvs?
3. I see that `RetargetingMethodSymbol.UnmanagedCallersOnlyAttributeData` retargets the `CallingConventionTypes`. Can we cover that?

Reply:
1. Sure, I can verify that.
2. Empty, yes. `null`, no, the only scenario that can happen is if the data is uninitialized.
3. That's what this test does :)

---
In reply to: [481428464](https://github.com/dotnet/roslyn/pull/47311#discussion_r481428464) [](ancestors = 481428464)